pkg/credentialprovider/azure/azure_credentials.go

```diff
@@ -301,3 +302,26 @@ func (a *acrProvider) parseACRLoginServerFromImage(image string) string {
 
 	return ""
 }
+
+// shouldUseCache decide whether to use cache according to image and cache content
+func shouldUseCache(image string, d credentialprovider.DockerConfig) bool {
```

Comment:
It's not clear to me why it is correct for the caching decision to use the image as an input. If this returns true, the results are cached and used for the next call to Provide(), regardless of the next image passed in.

Reply:
every time when `Provide(image string)` is invoked, it would use image to check whether `d credentialprovider.DockerConfig` contains server in `image`, since Azure Managed Identity uses different token per registry server, that's the reason why I worked out this PR.

Reply:
> Azure Managed Identity uses different token per registry server

if Provide is returning credentials that are only usable against the input image's registry, it doesn't seem like they should be cached for all subsequent calls to Provide, regardless of image (which is how CachingDockerConfigProvider currently works)

Reply:
I would say it depends on image, e.g. `foo.azurecr.io/image:ver`, `foo.azurecr.io/image2:ver`, these two images should use cache (otherwise it may cause more API to get registry token). In most cases, user would use one registry to store most of images, then we should use cache; while in other cases, if user use 2+ registries, we sometimes don't use cache.

Reply:
> I would say it depends on image, e.g. `foo.azurecr.io/image:ver`, `foo.azurecr.io/image2:ver`, these two images should use cache (otherwise it may cause one more API to get registry token).

The way this is currently implemented seems like a roundabout way to get per-image caching.

Would it make more sense to provide a cache key function that determines the cache key for a given image, orthogonal to ShouldCache? Something like:

```diff
 // Enabled implements dockerConfigProvider
@@ -94,16 +100,50 @@ func (d *CachingDockerConfigProvider) Provide(image string) DockerConfig {
        d.mu.Lock()
        defer d.mu.Unlock()
 
+       key := ""
+       if d.CacheKey != nil {
+               key = d.CacheKey(image)
+       }
+
        // If the cache hasn't expired, return our cache
-       if time.Now().Before(d.expiration) {
-               return d.cacheDockerConfig
+       if time.Now().Before(d.expiration[key]) {
+               return d.cacheDockerConfig[key]
        }
 
        klog.V(2).Infof("Refreshing cache for provider: %v", reflect.TypeOf(d.Provider).String())
        config := d.Provider.Provide(image)
+       now := time.Now()
        if d.ShouldCache == nil || d.ShouldCache(config) {
-               d.cacheDockerConfig = config
-               d.expiration = time.Now().Add(d.Lifetime)
+               if d.cacheDockerConfig == nil {
+                       d.cacheDockerConfig = map[string]DockerConfig{}
+               }
+               d.cacheDockerConfig[key] = config
+               if d.expiration == nil {
+                       d.expiration = map[string]time.Time{}
+               }
+               d.expiration[key] = now.Add(d.Lifetime)
        }
+
+       c.cleanupIfNeeded_locked()
+
        return config
 }
+
+// cleanupIfNeeded_locked removes expired credentials from the cache if needed.
+// must be called with mu already locked.
+func (c *CachingDockerConfigProvider) cleanupIfNeeded_locked() {
+       // skip cleanup if we have very few credentials or our cleanup time has not yet arrived
+       if len(d.cacheDockerConfig) < 100 || now.Before(d.nextCleanupTime) {
+               return
+       }
+
+       // remove expired credentials
+       for key := range d.cacheDockerConfig {
+               if now.Before(d.expiration[key]) {
+                       delete(d.expiration, key)
+                       delete(d.cacheDockerConfig, key)
+               }
+       }
+       // set our next cleanup time
+       d.nextCleanupTime = now.Add(d.Lifetime)
+}
```

Reply:
@liggitt thanks for the suggestion. There are two identities on Azure now: Service Principal(SP) and Managed Identity(MI), and we only want to make this change happen in MI since for SP, use cache is always ok. So based on your suggestion, we also need to add identity check here:
```
if time.Now().Before(d.expiration[key]) {
```

That would make the implementation too cloud provider specific, since this feature is going to retire in the future(use a separator credential provider for every cloud provider), I think it may be better to make this change as small as possible, otherwise in common code, we need to handle too cloud provider specific thing.

Let me know if you have any questions, or ping me in slack to make it quick, thanks!

Reply:
> we only want to make this change happen in MI

could that be determined by looking at `a.config.UseManagedIdentityExtension`? would a custom CacheKey impl like this work without adding anything cloud-specific to CachingDockerConfigProvider?

```go
func(image string) (cacheKey string) {
  if !a.config.UseManagedIdentityExtension {
    // not using managed identity, use the same credentials for all images
    return ""
  }
  // cache credentials per loginServer
  return a.parseACRLoginServerFromImage(image)
}
```

Reply:
that would make the implementation and logic very complicated, shall we go with this PR? Thanks.

Reply:
The current state of the PR seems fragile to me, for several reasons:
1.  It parses the image differently (splitting on / vs parseACRLoginServerFromImage) for purposes of determining if the returned credentials have an entry for the registry.
2. It pins to a credential set length to detect if managed identity is in use, instead of checking the credential provider config
3. It decreases cache hits when multiple registries are used, and it's not clear to me how impactful that is, performance-wise. 
4. It seems to use false-positive methods for determining if a given image is in a given domain (contains azureCRName)

Reply:
@liggitt thanks for the review. I moved `ShouldUseCache` func to provider interface, now the logic is more clear:
within one minute cache expiration time, if `d.provider.ShouldUseCache(image, config)` return false (only for Azure Managed Identity), it would merge the two cache maps. This PR has addressed point. 1-4, and only use one expiration date, make the logic more clear, PTAL, thanks!

Reply:
@liggitt could you review on this PR again? I have followed up with points 1-4, Thanks. 
the cache is per registry, not per image. e.g. if image is `foo.azurecr.io/image1:v1`, then cache would be `foo.azurecr.io`, when `foo.azurecr.io/image1:v2` comes, it hit the cache, while when `foo2.azurecr.io/image1:v1` comes cache missed. This PR would fix such cache missing issue. I think we can not do a per image cache, which may be too expensive.

Reply:
having a `ShouldUseCache(image, config) bool` method combined with a cache that does not differentiate by image does not make sense... 

> the cache is per registry, not per image

That is not reflected in a single boolean return value. I think the signature should be something like 
```
CacheKey(image string) string
```

I also think cache-related methods (e.g. CacheKey or ShouldUseCache) makes more sense as an option to CachingDockerConfigProvider, since it is only used by that and is not inherent to the provider